keep-runtime/src/workload.rs

```diff
@@ -0,0 +1,134 @@
+    ImportModuleNotFound(String),
+    /// import field not found
+    ImportFieldNotFound(String, String),
+    /// export not found
+    ExportNotFound,
+    /// call failed
+    CallFailed,
+    /// runtime error
+    RuntimeError,
+    /// I/O error
+    IoError(std::io::Error),
+}
+
+impl From<std::io::Error> for Error {
+    fn from(err: std::io::Error) -> Self {
+        Self::IoError(err)
+    }
+}
+
+/// Result type used throughout the library.
+pub type Result<T> = std::result::Result<T, Error>;
+
+/// Runs a WebAssembly workload.
+pub fn run<S: AsRef<[u8]>, T: std::borrow::Borrow<(S, S)>>(
+    bytes: impl AsRef<[u8]>,
+    args: impl IntoIterator<Item = S>,
+    envs: impl IntoIterator<Item = T>,
+) -> Result<Box<[wasmtime::Val]>> {
+    let engine = wasmtime::Engine::default();
+    let store = wasmtime::Store::new(&engine);
+
+    // Instantiate WASI
+    let mut builder = wasi_common::WasiCtxBuilder::new();
+    builder.args(args).envs(envs);
+    let ctx = builder.build().or(Err(Error::RuntimeError))?;
+    let wasi_snapshot_preview1 = wasmtime_wasi::Wasi::new(&store, ctx);
+
+    let instance = {
+        let module = wasmtime::Module::new(&store, bytes).or(Err(Error::RuntimeError))?;
+        let imports = module
+            .imports()
+            .map(|import| {
+                let module_name = import.module();
+                let field_name = import.name();
+                let export = match module_name {
+                    "wasi_snapshot_preview1" => Ok(wasi_snapshot_preview1.get_export(field_name)),
+                    _ => Err(Error::ImportModuleNotFound(module_name.to_string())),
+                }?;
+
+                if let Some(export) = export {
+                    Ok(export.clone().into())
+                } else {
+                    Err(Error::ImportFieldNotFound(
+                        module_name.to_string(),
+                        field_name.to_string(),
+                    ))
+                }
+            })
+            .collect::<Result<Vec<_>>>()?;
+
+        wasmtime::Instance::new(&module, &imports).or(Err(Error::RuntimeError))?
+    };
+
+    let function = if instance.exports().any(|export| export.name().is_empty()) {
+        // Launch the default command export.
+        instance.get_func("")
+    } else {
+        // If the module doesn't have a default command
+        // export, launch the _start function if one is
+        // present, as a compatibility measure.
+        instance.get_func("_start")
+    }
+    .ok_or(Error::ExportNotFound)?;
+
+    // Invoke the function.
+    function.call(Default::default()).or(Err(Error::CallFailed))
+}
+
+#[cfg(test)]
+pub(crate) mod test {
+    use crate::workload;
+
+    #[test]
+    fn workload_run_return_1() {
+        let path = std::path::Path::new("fixtures").join("return_1.wat");
+
+        let bytes = wat::parse_file(&path).unwrap();
+
+        let results: Vec<i32> = workload::run(&bytes, Vec::<String>::new(), &[])
```

Comment:
The code you've provided still performs an allocation (unless the optimizer removes it). We probably want something like this instead (which is more readable too):

```suggestion
        let args: [&str; 0] = [];
        let results: Vec<i32> = workload::run(&bytes, &args, &[])
```

Reply:
FYI: https://github.com/rust-lang/rfcs/blob/master/text/0803-type-ascription.md

Reply:
Err. That's the old one. Here's the new one: https://github.com/rust-lang/rfcs/pull/2522

Reply:
This might be better:

```suggestion
        use std::iter::empty();
        let results: Vec<i32> = workload::run(&bytes, empty::<&str>(), &[])
```

Reply:
That helps; thanks!